File: services/rust/rocket/src/routes/handlers/patch.rs
use crate::routes::{
    errors::AppError,
    extractors::{AppJson, AppPath},
    guards::ChaosGuard,
};
use app::{state::AppState, task::PatchedTask};
use rocket::{State, http::Status, patch, serde::json::Json};
use serde_json::{Value, json};
use std::sync::{Arc, Mutex};
use tracing::{error, info, instrument, warn};

#[patch("/<id>", data = "<body>")]
#[instrument(skip_all)]
pub async fn patch_handler(
    _guard: ChaosGuard,
    id: AppPath,
    state: &State<Arc<Mutex<AppState>>>,
    body: AppJson<PatchedTask>,
) -> (Status, Json<Value>) {
    let id = id.into_inner();

    if let Ok(mut state) = state.lock() {
        if let Some(task) = state.tasks.get_mut(&id) {
            let previous_operation = task.operation.clone();
            task.operation = body.into_inner().operation;

            info!(
                %id,
                secret = task.secret.len(),
                from_operation = previous_operation.to_string(),
                to_operation = task.operation.to_string(),
                method = "PATCH",
                "Patched task"
            );

            return (Status::Ok, Json(json!(task.to_owned())));
        } else {
            drop(state);
            warn!(%id, method = "PATCH", "Task not found");
            return AppError::TaskNotFound.into_response();
        }
    }

    error!(%id, method = "PATCH", "Poisoned lock");

    AppError::Internal.into_response()
}

File: services/go/fiber/routes/handlers/patch.go
package handlers

import (
	"app"
	"fiber/routes"
	"log/slog"
	"net/http"
	"strings"

	"github.com/gofiber/fiber/v3"
)

func PatchHandler(state *app.AppState) fiber.Handler {
	return func(c fiber.Ctx) error {
		id, err := ParseID(c)
		if err != nil {
			return err
		}

		patchedTask, err := ParsePatchedTask(c, c.Body())
		if err != nil {
			return err
		}

		state.Mu.Lock()
		task, ok := state.Tasks[id]

		if !ok {
			state.Mu.Unlock()
			slog.Warn("Task not found", "method", c.Method(), "path", c.Path(), "id", id.String())
			return routes.AppErrors.TaskNotFound.Error(c)
		}

		previousOperation := task.Operation
		task.Operation = patchedTask.Operation
		state.Tasks[id] = task
		state.Mu.Unlock()

		slog.Info(
			"Patched task",
			"id", id.String(),
			"secret", len(task.Secret),
			"from_operation", strings.ToLower(string(previousOperation)),
			"to_operation", strings.ToLower(string(patchedTask.Operation)),
			"method", c.Method(),
		)

		return c.Status(http.StatusOK).JSON(task)
	}
}
